Add boiler-associated attributes from EIA 860 6.2 EnvrEquip tables to ETL

e-belfer opened the pull request:
This issue scopes out a subset of work identified in the issues https://github.com/catalyst-cooperative/pudl/issues/1189, https://github.com/catalyst-cooperative/pudl/issues/1162, https://github.com/catalyst-cooperative/pudl/issues/889.

EIA 860 Schedule 6 (Environmental Equipment and Environmental Association) reports data on the environmental quality control equipment, investments, regulatory standards and control strategies associated with particular boilers. Much of this information will need to be organized outside of our existing EIA860 data structure, as multiple pieces of control equipment may be associated with multiple boilers. More immediately, however, data associated with a boiler ID can be integrated into the existing harvesting process. To bring boiler-associated attributes from these two datasets into PUDL, this PR incorporates two new tables into `boilers_eia860` and `boilers_entity_eia`: `emissions_control_strategies` and `boiler_info` (from 6.2 Environmental Equipment). Since all columns in both tables match 1:1 with boilers, they are entirely harvested into these two tables and do not separately appear in the final ETL output.

Notable decisions:
- Dropped `mercury_emissions_control` boolean (old boolean column whose information is contained in other columns).
- Converted the many boolean mercury equipment categories (2009-2012) from `emissions_control_strategies` into codes in the `mercury_existing_strategy` columns. Newer data has three of these columns but as many as six pieces of equipment are found in older columns, so there are now 6 of these columns.
- Where columns get less specific over time, I've for now merged the older columns to match the latest set of codes. The strategy codes (e.g. `pm_control_out_of_compliance_strategy_1`) get much less granular over time. This is most notable for the SO2 codes, with the addition of an "Use flue gas desulfurization unit or other SO2 control process (specify the specific type of equipment in Schedule 6A)" category for all equipment-based emission control strategies that likely links to one or more fields in the environmental equipment dataset. For now I've included all the codes that aren't easily mapped to the present set of codes so as to not lose this data. Eventually we'll want to more thoroughly reconfigure these codes to be consistent across time when the equipment data is integrated.
- Where columns get more specific over time, I've for now added the older more generic columns into the encoding tables. 
- Renamed `operating_date`, `retirement_date` and related columns to refer specifically to generators and make clear the distinction between generator dates and boiler dates. This needed to be updated in many places, including tests, which is why this PR has so many file changes!
- Added `boiler_manufacturer` and `nox_control_manufacturer` as a name in addition to its code column, for legibility.
- Updated default `.yml` files to include boilers_eia860 table.

Specific questions for reviewers:
- Presumably the new tables need new/updated validation tests? How do we want to deal with this in this interim step in the ETL? In the future there will need to be substantial additional validation tests for the boilers tables, but I imagine it's worth waiting until we integrate the rest of the data and finalize the table's form a bit more before we write these?

PR to-dos:
- [x] Confirm static vs annual attributes
- [x] Reconfigure `boilers_entity_eia`: deal with the `prime_mover_code`
- [x] Write `boilers_eia860` output table
- [x] Write data description for tables
- [x] Update naming conventions
- [x] Add boiler output table to eia validation tests
- [x] prime_mover_code has a m:1 relationship to boilers. Remove it from the boiler harvest tables.
- [x] Fix boil_eia860 output table (utility ID is currently null)

Future work:
- Streamline the control strategy codes
- Some of these columns likely split off and moved into the data tables associated with specific equipment. As we incorporate more equipment data, we may want to rethink where columns are ultimately living.

# PR Checklist

- [x] Merge the most recent version of the branch you are merging into (probably `dev`).
- [x] All CI checks are passing. [Run tests locally to debug failures](https://catalystcoop-pudl.readthedocs.io/en/latest/dev/testing.html#running-tests-with-tox)
- [x] Make sure you've included good docstrings.
- [x] For major data coverage & analysis changes, [run data validation tests](https://catalystcoop-pudl.readthedocs.io/en/latest/dev/testing.html#data-validation)
- [x] Include unit tests for new functions and classes.
- [x] Defensive data quality/sanity checks in analyses & data processing functions.
- [x] Update the [release notes](https://catalystcoop-pudl.readthedocs.io/en/latest/release_notes.html) and reference the PR and related issues.
- [x] Do your own explanatory review of the PR to help the reviewer understand what's going on and identify issues preemptively.

zaneselvans commented on src/pudl/metadata/resources/__init__.py:
```diff
@@ -201,8 +201,84 @@
-        "static_cols": ["prime_mover_code"],
-        "annual_cols": [],
+        "static_cols": [
+            "prime_mover_code",
+            "state",
+            "county",
+            "operating_date",
+            "retirement_date",
```
I think all of these column names appear in other tables too, which will probably break the harvesting process. I think each attribute is expected to be uniquely associated with a single entity. The `state` and `county` (and other location fields) are part of `plants_entity_eia`, while `prime_mover_code` is associated with `generators_eia860` I could see there being separate operating & retirement dates for generators and boilers, but in that case they'll need to have unique names, e.g. `generator_operating_date` vs. `boiler_operating_date`, otherwise they'll be confused by the harvesting process.

cmgosnell replied:
I know i'm jumping in later here but i'm not sure this is totally true about confusing the harvester. the harvesting would be confused if a table has **both** `boiler_id` and `generator_id` **and** these columns we want to harvest.

The harvester loops through these entities, finds the tables with the `id_cols` then rips all instances of the static and annual cols out of their original tables (unless those cols are in `keep_cols`). But from what I see of these new tables that your pulling in now, there aren't table with both `boiler_id` and `generator_id` in them. Although I could be missing it!

I'm personally partial to having the same column names for things even when they are referring to seperate entities if they generally mean the same thing.

e-belfer replied:
Jumping slightly ahead, when we integrate `emissions_control_equipment` we'll get another set of operating and retirement date columns, this time associated with the control equipment. This table will have the boiler ID in it and it's possible we'd want to aggregate some of the columns (e.g. total cost of control equipment) to add to the harvest (the rest have an m:1 relationship to boilers). Between this and the actual different meaning of "operating_date" between generators (actual operating date) and boilers (planned or actual), I wonder if distinguishing between source is the most useful here?

cmgosnell replied:
yeah this is a good point. I think I am generally partial to have things named the same way but if it introduces lots of downstream confusion then I feel good about these more specific names.

zaneselvans commented on src/pudl/metadata/codes.py:
```diff
@@ -53,6 +134,424 @@
         },
         "ignored_codes": [],
     },
+    "envr_equipment_manufacturer_codes_eia": {
```
Wacky that they decided to use codes for the manufacturers instead of names.

e-belfer replied:
Indeed. We could decide to just skip the codes and replace them with the manufacturer name directly, might be more usable.

zaneselvans commented on src/pudl/metadata/codes.py:
```diff
@@ -53,6 +134,424 @@
+        "code_fixes": {
+            "AF": "VF",
+            "CF": "TF",
+            "FF": "WF",
+            "OF": "WF",
+            "RF": "WF",
+            "SF": "WF",
```
Did these codes get more or less granular over time? If they got more granular over time, does it seem like we could backfill with the more detailed codes?

e-belfer replied:
The codes get less granular over time.

zaneselvans replied:
I might be misunderstanding your summary from the top level PR comment, but it seemed like in some cases you were merging codes, and in others you were keeping both the generic and granular codes. If that's true, what's your thinking there? Are you just trying to make sure that the data reflects the most recent reporting structure?

e-belfer replied:
Where codes are getting less granular over time (i.e. here, where the newer code WF includes older codes FF, OF, RF, SF), I have merged them to reflect the most recent reporting structure. Where older data is less specific (e.g. just specifies a dry scrubber) and I'm not able to clearly identify what more specific subcategory the record could be matched to, I add the older more generic categories to the dictionary and make a note in the field metadata.

zaneselvans commented on src/pudl/metadata/codes.py:
```diff
@@ -53,6 +134,424 @@
+        "code_fixes": {},
+        "ignored_codes": [],
+    },
+    "pm_unit_codes_eia": {
```
Wow these are some wild units.

e-belfer replied:
Indeed...

zaneselvans commented on src/pudl/metadata/codes.py:
```diff
@@ -1480,4 +1979,144 @@
+        "df": pd.DataFrame(
+            columns=["code", "label", "description"],
+            data=[
+                ("BO", "out_of_serive", "Burner out of service."),
```
Typo? `out_of_serive` => `out_of_service`

zaneselvans reviewed (commented):
I suspect the harvesting step is breaking because the same column names are listed as being associated with more than one entity (generators, plants, boilers).

e-belfer commented:
Some outstanding design decisions to resolve:
- The strategy codes (e.g. `pm_control_out_of_compliance_strategy_1`) get much less granular over time. This is most notable for the SO2 codes, with the addition of an `"Use flue gas desulfurization unit or other SO2 control process (specify the specific type of equipment in Schedule 6A)"` category for all equipment-based emission control strategies that likely links to one or more fields in the environmental equipment dataset. For now I've included all the codes that aren't easily mapped to the present set of codes so as to not lose this data. Eventually we'll want to more thoroughly reconfigure these codes to be consistent across time when the equipment data is integrated, though I'm not intuitively sure how best to go about this.
- the `energy_source_code` fields are provided for both generators (1-6) and boilers (1-8). Are these two separate sets of codes? If not, what table is the best place for them to ultimately live, and what's the best way to move them between entities?
- Do the allocations between annual and static columns make sense?

zaneselvans commented:
I think preserving the strategy codes for now makes sense, until we can figure out whether that information is available in the other environmental equipment data. One thing we've done in a few cases, like the `energy_sources_eia` table, is store broader categories as additional columns, so the information about what codes map to which categories is available in the DB at least.

It's interesting that `energy_source_code` is getting reported both in association with the generators and the boilers. I wonder if there's a lot of diversity within individual generation units (collections of connected boilers and generators), or if the `energy_source_code` should really pertain to the entire unit, rather than the individual boilers and generators.

With generators, a lot of the `energy_source_code` values don't necessarily describe fuels, but I guess with boilers they really will all be fuels.  If there are up to 8 values being reported `primary_fuel` seems a little misleading. Maybe `boiler_fuel_code_N`?

In the past we've often made the mistake of identifying very slowly changing attributes as "static" and then later had to revise them to be annual, because we incorrectly assumed that the changes were reporting errors initially. It's been helpful to find cases in which a value changes and research it in particular, to see if the changes can be real. For example, there are rare cases in which the `prime_mover_code` associated with a particular generator actually does change, but we were only able to confirm that with some googling of individual generators and their histories.

e-belfer commented:
Still to do:
- update `primary_fuel_n` code name to `boiler_fuel_n`
- check for 0.0's that should be nans
- figure out why ci-test is breaking
- check all static codes to confirm "static-ness"

Re: `energy_source_code`: these originally have different names in the EIA dataset. I'd been following Alex's lead to make the naming conventions consistent but one's column definition explicitly refers to the generator and one to the boiler. If we want to harvest these to one unit-level code I'd suggest making a new set of columns to do so?

e-belfer commented:
Next steps:
- add encoding tables for all new columns from `emissions_control_strategies`
- If any other 'proposed/existing' strategy columns report emissions control strategies, add them to the main proposed/existing control strategies columns.
- run through the standard cleaning checks for new columns
- confirm that firing_rate cols should remain static rather than annual

e-belfer commented on src/pudl/transform/eia860.py:
```diff
@@ -579,6 +579,102 @@ def _make_phone_number(col1, col2, col3):
+    zero_columns_to_fix = [
+        "firing_rate_using_coal_tons_per_hour",
+        "firing_rate_using_gas_mcf_per_hour",
+        "firing_rate_using_oil_bbls_per_hour",
```
Just a note: Using __eq__ here since flake8 took issue with ==True.

zaneselvans commented on src/pudl/transform/eia860.py:
```diff
@@ -603,25 +606,154 @@ def boilers(eia860_dfs, eia860_transformed_dfs):
+            )
         )
 
+    # 2009-2012 data uses boolean columns for mercury equipment that
+    # later are converted to strategy codes. Here we convert them manually.
+
+    mercury_boolean_cols = [
+        "ACI",
+        "BH",
+        "DS",
+        "EP",
+        "FGD",
+        "LIJ",
+        "WS",
+        "BS",
+        "BP",
+        "BR",
+        "EC",
+        "EH",
+        "EK",
+        "EW",
+        "OT",
+    ]
```
Can you expand on the comment here explaining what's going on with the mercury boolean columns? It sounds like there was a structural change in how the old vs. new data is being reported, and you're reshaping the old data to conform to the new structure, which is N columns that contain values from the same set of codes.

Not something to tackle right now, but it looks like there are a growing number of collections of numbered coded columns like this, and I wonder if we might want to think about how to best normalize them in the DB. Having a single column that can contain one of several different codes makes sense (e.g. with prime mover), but when we need to be able to record up to 6 or 8 different values, I wonder if boolean columns, with one per potential code, might make more sense. Or we could do a bad thing and put a list of codes in a single column.  Is there information in the cardinality of the columns? I.e. is the emissions control strategy in the `_1` column more important than the one in the `_2` column?

How do we imagine folks using these columns in analysis? What structure will make selecting on the basis of the codes present easiest? If you wanted to use them as features for an ML algorithm (say, to estimate the expected emissions for a unit that doesn't report its emissions directly in the EPA CEMS) how would you want to arrange them? Is it ever possible to have more than 6 mercury control strategies on a single unit?

e-belfer replied:
Yes, I agree. I think the best course of action will be clearer once we decide how we want to integrate the respective equipment data columns. It could be that we just replace all of these columns with a link to the respective equipment IDs and keep information about equipment types in those dataframes, or that we want to pursue some kind of boolean reorganization. Currently there are loads of columns that differentiate between different contexts (current/planned, what regulation is being met, etc.), so we could also pursue recoding in other columns (e.g. most string applicable regulation) to reduce the number of columns functioning here. There is no information about cardinality in the columns explicitly specified in the layout files or the survey, so I don't think we can presume that here.

zaneselvans commented on src/pudl/package_data/eia860/column_maps/emission_control_strategies.csv:
```diff
@@ -52,20 +52,20 @@ steam_plant_type_code,,,,,,steam_plant_type,steam_plant_type,steam_plant_type,st
-mercury_emission_electrostatic_precipitator_hot_side_without_flue_gas_conditioning,,,mercury_emission_ek,mercury_emission_electrostatic_precipitator_hot_side_without_flue_gas_conditioning,,,,,,,,,
-mercury_emission_electrostatic_precipitator_cold_side_without_flue_gas_conditioning,,,mercury_emission_ew,mercury_emission_electrostatic_precipitator_cold_side_without_flue_gas_conditioning,,,,,,,,,
-mercury_emission_other,mercury_emission_ot,mercury_emission_ot,mercury_emission_ot,mercury_emission_other,,,,,,,,,
+ACI,mercury_emission_acj,mercury_emission_acj,mercury_emission_acj,mercury_emission_activated_carbon_injection,,,,,,,,,
+BH,mercury_emission_bh,mercury_emission_bh,mercury_emission_bh,mercury_emission_baghouse,,,,,,,,,
+DS,mercury_emission_ds,mercury_emission_ds,mercury_emission_ds,mercury_emission_dry_scrubber,,,,,,,,,
+EP,mercury_emission_ep,mercury_emission_ep,mercury_emission_ep,mercury_emission_electrostatic_precipitator,,,,,,,,,
+FGD,mercury_emission_fgd,mercury_emission_fgd,mercury_emission_fgd,mercury_emission_flue_gas_desulfurization,,,,,,,,,
+LIJ,mercury_emission_li,mercury_emission_li,mercury_emission_li,mercury_emission_lime_injecton,,,,,,,,,
+WS,mercury_emission_ws,mercury_emission_ws,mercury_emission_ws,mercury_emission_wet_scrubber,,,,,,,,,
+BS,,,mercury_emission_bs,mercury_emission_baghouse_shake_and_deflate,,,,,,,,,
+BP,,,mercury_emission_bp,mercury_emission_baghouse_pulse,,,,,,,,,
+BR,,,mercury_emission_br,mercury_emission_baghouse_reverse_air,,,,,,,,,
+EC,,,mercury_emission_ec,mercury_emission_electrostatic_precipitator_cold_side_with_flue_gas_conditioning,,,,,,,,,
+EH,,,mercury_emission_eh,mercury_emission_electrostatic_precipitator_hot_side_with_flue_gas_conditioning,,,,,,,,,
+EK,,,mercury_emission_ek,mercury_emission_electrostatic_precipitator_hot_side_without_flue_gas_conditioning,,,,,,,,,
+EW,,,mercury_emission_ew,mercury_emission_electrostatic_precipitator_cold_side_without_flue_gas_conditioning,,,,,,,,,
+OT,mercury_emission_ot,mercury_emission_ot,mercury_emission_ot,mercury_emission_other,,,,,,,,,
```
Are these codes selected to match the ones that show up in later years? `ACI` and `LIJ` seem to be kind of inconsistent (but if that's the way the latter codes are organized, so be it!)

e-belfer replied:
Yes, ACI and LIJ are re-coded to match the dictionary provided in the latest 860 data.

e-belfer commented on src/pudl/metadata/fields.py:
```diff
@@ -884,6 +924,18 @@
+        "type": "date",
+        "description": "Date the generator began commercial operation.",
+    },
+    "generator_operating_year": {
+        "type": "integer",
+        "description": "Year a generator went into service.",
+    },
```
These both show up in the plant_parts table. I'm not sure what the need is for both columns, I'm just updating what was in the dictionary initially.

zaneselvans replied:
Okay, then I think it's something specific to the PPE output table, which also needs to get typed, and it doesn't end up in the DB.

e-belfer commented:
The following validation tests are failing and need to be updated:
```
FAILED test/validate/eia_test.py::test_minmax_rows[eia_raw-bf_eia923-1415232-1415232-118550] - ValueError: bf_eia923: found 1415328 rows, expected 1415232. Of...
FAILED test/validate/eia_test.py::test_minmax_rows[eia_raw-frc_eia923-596855-244403-24064] - ValueError: frc_eia923: found 597000 rows, expected 596855. Off...
FAILED test/validate/eia_test.py::test_minmax_rows[eia_raw-gf_eia923-2687321-2687321-230147] - ValueError: gf_eia923: found 2687345 rows, expected 2687321. Of...
FAILED test/validate/eia_test.py::test_minmax_rows[eia_raw-gf_nonuclear_eia923-2671268-2671268-228804] - ValueError: gf_nonuclear_eia923: found 2671292 rows, expected 2...
FAILED test/validate/eia_test.py::test_minmax_rows[eia_raw-plants_eia860-185551-185551-185551] - ValueError: plants_eia860: found 185553 rows, expected 185551. ...
FAILED test/validate/eia_test.py::test_minmax_rows[eia_raw-pu_eia860-184740-184740-184740] - ValueError: pu_eia860: found 184743 rows, expected 184740. Off ...
FAILED test/validate/eia_test.py::test_minmax_rows[eia_raw-utils_eia860-119365-119365-119365] - ValueError: utils_eia860: found 119366 rows, expected 119365. O...
FAILED test/validate/eia_test.py::test_minmax_rows[eia_annual-bf_eia923-1415232-1415232-118550] - ValueError: bf_eia923: found 118558 rows, expected 118550. Off ...
FAILED test/validate/eia_test.py::test_minmax_rows[eia_annual-frc_eia923-596855-244403-24064] - ValueError: frc_eia923: found 24065 rows, expected 24064. Off b...
FAILED test/validate/eia_test.py::test_minmax_rows[eia_annual-gen_eia923-None-5171930-432657] - ValueError: gen_eia923: found 432570 rows, expected 432657. Off...
FAILED test/validate/eia_test.py::test_minmax_rows[eia_annual-gf_eia923-2687321-2687321-230147] - ValueError: gf_eia923: found 230149 rows, expected 230147. Off ...
FAILED test/validate/eia_test.py::test_minmax_rows[eia_annual-gf_nonuclear_eia923-2671268-2671268-228804] - ValueError: gf_nonuclear_eia923: found 228806 rows, expected 22...
FAILED test/validate/eia_test.py::test_minmax_rows[eia_annual-plants_eia860-185551-185551-185551] - ValueError: plants_eia860: found 185553 rows, expected 185551. ...
FAILED test/validate/eia_test.py::test_minmax_rows[eia_annual-pu_eia860-184740-184740-184740] - ValueError: pu_eia860: found 184743 rows, expected 184740. Off ...
FAILED test/validate/eia_test.py::test_minmax_rows[eia_annual-utils_eia860-119365-119365-119365] - ValueError: utils_eia860: found 119366 rows, expected 119365. O...
FAILED test/validate/mcoe_test.py::test_minmax_rows_mcoe[eia_annual-hr_by_unit-361963-30305] - ValueError: hr_by_unit: found 30309 rows, expected 30305. Off b...
FAILED test/validate/mcoe_test.py::test_minmax_rows_mcoe[eia_annual-hr_by_gen-554617-46366] - ValueError: hr_by_gen: found 46370 rows, expected 46366. Off by...
FAILED test/validate/mcoe_test.py::test_minmax_rows_mcoe[eia_annual-fuel_cost-554617-46366] - ValueError: fuel_cost: found 46370 rows, expected 46366. Off by...
FAILED test/validate/mcoe_test.py::test_minmax_rows_mcoe[eia_annual-capacity_factor-5171930-432657] - ValueError: capacity_factor: found 432570 rows, expected 432657...
FAILED test/validate/mcoe_test.py::test_minmax_rows_mcoe[eia_annual-mcoe-5172314-432689] - ValueError: mcoe: found 432602 rows, expected 432689. Off by -0...
FAILED test/validate/eia_test.py::test_minmax_rows[eia_monthly-bf_eia923-1415232-1415232-118550] - ValueError: bf_eia923: found 1415328 rows, expected 1415232. Of...
FAILED test/validate/eia_test.py::test_minmax_rows[eia_monthly-frc_eia923-596855-244403-24064] - ValueError: frc_eia923: found 244415 rows, expected 244403. Off...
FAILED test/validate/eia_test.py::test_minmax_rows[eia_monthly-gen_eia923-None-5171930-432657] - ValueError: gen_eia923: found 5171497 rows, expected 5171930. O...
FAILED test/validate/eia_test.py::test_minmax_rows[eia_monthly-gf_eia923-2687321-2687321-230147] - ValueError: gf_eia923: found 2687345 rows, expected 2687321. Of...
FAILED test/validate/eia_test.py::test_minmax_rows[eia_monthly-gf_nonuclear_eia923-2671268-2671268-228804] - ValueError: gf_nonuclear_eia923: found 2671292 rows, expected 2...
FAILED test/validate/eia_test.py::test_minmax_rows[eia_monthly-plants_eia860-185551-185551-185551] - ValueError: plants_eia860: found 185553 rows, expected 185551. ...
FAILED test/validate/eia_test.py::test_minmax_rows[eia_monthly-pu_eia860-184740-184740-184740] - ValueError: pu_eia860: found 184743 rows, expected 184740. Off ...
FAILED test/validate/eia_test.py::test_minmax_rows[eia_monthly-utils_eia860-119365-119365-119365] - ValueError: utils_eia860: found 119366 rows, expected 119365. O...
FAILED test/validate/mcoe_test.py::test_minmax_rows_mcoe[eia_monthly-hr_by_unit-361963-30305] - ValueError: hr_by_unit: found 362011 rows, expected 361963. Off...
FAILED test/validate/mcoe_test.py::test_minmax_rows_mcoe[eia_monthly-hr_by_gen-554617-46366] - ValueError: hr_by_gen: found 554665 rows, expected 554617. Off ...
FAILED test/validate/mcoe_test.py::test_minmax_rows_mcoe[eia_monthly-fuel_cost-554617-46366] - ValueError: fuel_cost: found 554665 rows, expected 554617. Off ...
FAILED test/validate/mcoe_test.py::test_minmax_rows_mcoe[eia_monthly-capacity_factor-5171930-432657] - ValueError: capacity_factor: found 5171497 rows, expected 51719...
FAILED test/validate/mcoe_test.py::test_minmax_rows_mcoe[eia_monthly-mcoe-5172314-432689] - ValueError: mcoe: found 5171881 rows, expected 5172314. Off by ...
```

zaneselvans commented on src/pudl/transform/eia.py:
```diff
@@ -327,13 +327,12 @@ def _compile_all_entity_records(entity, eia_transformed_dfs):
     # most columns become objects (ack!), so assign types
     compiled_df = apply_pudl_dtypes(compiled_df, group="eia")
-    # encode the compiled options! (except for the boilers bc there is no annual table)
-    if entity != "boilers":
```
Yay! Goodbye silly special case. 🎉

zaneselvans reviewed (changes requested):
This is so much new data! You've done a great job navigating this messy machine.  I've tagged @cmgosnell in a few questions since she has more context about where some of these columns are ending up or will end up.

* I think the listing of `prime_mover_code` in both boiler & generator entities will cause problems and we may need to rename.
* I'm concerned about how confident we are in the static nature of some of the boiler attributes (including `prime_mover_code` given that it's not static for generators).
* But mostly comments about naming conventions.

zaneselvans reviewed (commented):
All these changes in expected row numbers look very reasonable. I do think we should spot check them just to understand why they're changing at all. I suspect that all the changes are flowing from the entity tables (generator, boiler, plant, utility, both annual and static), so reading the tables directly from the DB and diffing them as dataframes (after setting the index to be their PK and sorting) should show us what's up without needing to look at all the data tables themselves.

We should also add a `boilers_eia860` output table to the `PudlTabl` object that's analogous to `gens_eia860`

e-belfer commented on src/pudl/metadata/resources/eia.py:
```diff
@@ -19,7 +19,21 @@
+                "boiler_operating_date",
+                "boiler_manufacturer",
+                "boiler_manufacturer_code",
+                "firing_rate_using_coal_tons_per_hour",
+                "firing_rate_using_oil_bbls_per_hour",
+                "firing_rate_using_gas_mcf_per_hour",
+                "firing_rate_using_other_fuels",
+                "waste_heat_input_mmbtu_per_hour",
+                "wet_dry_bottom",
+                "fly_ash_reinjection",
```
**Move to annual:**

Operating date: 
482 boilers from 306 plants have more than one date reported. 
- There's probably some mix of junk data and real changes here. 
- For 181 of these records, the changes in operating date that occur are changes to _planned_ operating dates (i.e., the operating date is ahead of the reporting date). Otherwise, anecdotally, these changes seem to be between sets of dates that far predate the reporting period (e.g. 1964 to 1979).
- 192 boilers have a date that changes more than a year from the first reported record to the last. (E.g., approximately 40% of these changes are persistent and substantial).
- 31 boilers have first and last dates which are identical, but one or more dates in the middle changes.
- For some very small proportion of boilers, the change in operating date is accompanied by a change in other design characteristics: 9 boilers (change in coal firing rate), 6 (change in oil firing rate), 33 (change in gas firing rate).

Hypotheses for changes: 1) plants are changing and/or mixing up the boiler IDs to which they assign equipment (e.g. renaming boiler 1 as boiler 3); 2) plants are reporting a suspected operating date for a boiler not yet in service, which is changing; 3) because the question is phrased as "in-service date", there may be a lack of clarity about whether to report the year the boiler first began operating, or the year it resumed operating after going offline for service, and practices may have changed over time; 4) plants may be substantially replacing or upgrading boilers.

Wet/dry bottoms:
Wet/dry bottom conversions are technically possible. We see this in 70 boilers, and value changes persist for multiple years.

Fly ash reinjection:
The vast majority of conversions here occur in concert with changes in other columns and persist (firing rate, turndown ratio, etc.), suggesting these reflect real design/operational changes in the boiler (or changes in what boiler is associated with the boiler ID, at least).

Firing rate changes:
As above, these seem to generally correspond to changes in other operational columns (e.g. turndown ratio).

Manufacturer / manufacturer code:
- 12 boilers from 4 plants change. 7 of these changes are from Alstom to "other, specify in schedule 7", two are from "General Electric" to "Nooter Erickson" in the same plant. All of the changes occur between 2009 and 2010. Arguably, this is the only column with a clear argument for keeping in static, but if it is the only one then perhaps it's worth just keeping all cols in the `boilers_eia860` table.

zaneselvans replied:
In my experience when a boiler or generator is replaced or converted to a new fuel (like coal steam generators getting repowered with gas) they **usually** change the ID, but it seems like there no universal patterns in their reporting, so I wouldn't be surprised if there are real exceptions.

* `planned_operating_date` seems like it could be quite changeable due to construction delays, and would make sense to keep in the annual table.
* Actual `operating_date` seems like it really should be constant, notwithstanding boilers that get replaced but keep an old ID.  I suspect that having multiple `operating_date` values on a single piece of infrastructure could cause some problems downstream, where we're using (generator) `operating_date` as a feature in the record linkage connecting FERC and EIA data. @cmgosnell what do you think?

For the physically descriptive attributes like wet/dry bottoms, fly ash reinjection, firing rate changes, I think if there's ambiguity, it's probably safest to make it a (very slowly changing) annual attribute, so we're preserving the reported information.

Not sure what to do about manufacturer. Like operating data, that really seems like it should be a fixed attribute conceptually, and 12 boilers out of... thousands? Tens of thousands? Not much to lose there.

e-belfer replied:
There is no `planned_operating_date` column for boiler data - the survey asks for "actual or proposed operating date" in one entry, so these are reported in the same column. If we wanted to split them out we would have to compare this against the reporting date and make a new column here.

I will say that the majority of changes in date (~300) are multi-year changes (e.g. a boiler will report 1964 from 2009-2013, and 1976 from 2014-present). All the hypotheses I've been able to come up with for why this would happen suggest that boiler ID is not a universally reliable attribute connecting boilers within all plants across years, but perhaps I'm missing something. 

Options:
1) Move everything to annual table.
2) Only keep manufacturer in the static table.
3) Keep manufacturer and operating date in the static table, creating about 300-400 NA records for operating date
4) Make a new planned operating date column comparing against the record date, keep that in the annual table and manufacturer and operating date in the static table, creating about 200-300 NA records for operating date.

e-belfer replied:
Arguments for leaving boiler manufacturer data in static make sense to me, so let's go for option 2 here.

e-belfer commented on src/pudl/metadata/resources/eia.py:
```diff
@@ -19,7 +19,21 @@
+                "waste_heat_input_mmbtu_per_hour",
+                "wet_dry_bottom",
+                "fly_ash_reinjection",
+                "prime_mover_code",
```
Re: naming. The shared name is not affected by the harvest because the boiler table it originates from doesn't include a generator ID, so we could _technically_ get away with not changing the names but I agree that from a clarity perspective renaming both would be useful.

 In `boiler_fuel_eia923` table from which the boiler `prime_mover_code` originates, one boiler can have more than one `prime_mover_code` associated with it in any given year (one record per plant, boiler and energy source code). Given this information, presumably it wouldn't make sense to harvest this record into either a static or annual table which expects 1:1 boiler attributes, since we'd lose information on 1:m records?

zaneselvans replied:
Yeah, that definitely sounds correct. What kind of situations do you find multiple prime movers attached to one boiler? Is it in combined cycle natural gas plants?  What's the right way to deal with this? Should we have more than one prime mover in association with each boiler? Several columns like we have for `energy_source_code`?

Does EIA's documentation provide any insight into the meaning of the single prime mover that's associated with the boiler in the `boiler_info` table?  Or is the `prime_mover_code` only coming from the `boiler_fuel_eia923` data table?  If it only appears in the data tables, and it doesn't represent an actual attribute of the boiler, then maybe it shouldn't be getting harvested at all. Again @cmgosnell may have more informed thoughts on this.

e-belfer replied:
It's only coming from `boiler_fuel_eia923`. I agree - in the latest iteration (not pushed yet but will be shortly) I've pulled this out of all the entity tables.

e-belfer replied:
In the end we're not harvesting prime_mover, so this question is no longer relevant.

cmgosnell replied:
i'm late to this party but it definitely seems to make sense to not harvest `prime_mover_code` for boilers!

zaneselvans commented on src/pudl/metadata/resources/eia.py:
```diff
@@ -39,6 +53,247 @@
         "etl_group": "static_eia",
         "field_namespace": "eia",
     },
+    "boiler_status_codes_eia": {
```
I think that sounds like a good approach.

zaneselvans commented on src/pudl/metadata/resources/__init__.py:
```diff
@@ -199,28 +193,33 @@
+            "wet_dry_bottom",
+            "fly_ash_reinjection",
+            "hrsg",
+            "max_steam_flow_thousand_pounds_per_hour",
```
There are some other thousands of pounds per hour (and pounds fields) if you search for `lbs` in `pudl.metadata.fields`. It looks like in general we've used the abbreviation `lbs` for pounds. We should probably use the same naming for all the thousands of pounds per hour fields, but I don't feel strongly about what it is.

zaneselvans commented on src/pudl/metadata/resources/eia.py:
```diff
@@ -17,22 +17,14 @@
         "field_namespace": "eia",
     },
     "boilers_entity_eia": {
+        # Note: This table is currently empty but we keep it for use in the crosswalk.
```
Is this still true?

zaneselvans commented on src/pudl/output/eia860.py:
```diff
@@ -530,6 +530,124 @@ def generators_eia860(
+    start_date=None,
+    end_date=None,
+    unit_ids: bool = False,
+    fill_tech_desc: bool = True,
```
I think `fill_tech_desc` only applies to generators.

zaneselvans commented on test/validate/eia_test.py:
```diff
@@ -40,18 +41,19 @@ def test_no_null_cols_eia(pudl_out_eia, live_dbs, cols, df_name):
         ("bga_eia860", 129_869, 129_869, 129_869),
-        ("frc_eia923", 596_855, 244_403, 24_064),
-        ("gen_eia923", None, 5_171_930, 432_657),
+        ("boil_eia860", 74_146, 74_146, 74_146),
```
Wow, that's a lot fewer than for the generators. I guess that's because there are a bunch of generators that don't burn anything and so don't have boilers.

e-belfer replied:
It seems roughly proportional to the bga table, but yes it is only applicable to this subset of the generators:

> For plants with a total steam-electric nameplate capacity of 10 MW or greater:
> Complete SCHEDULE 6, Part A for existing and planned boilers and associated equipment that serve
> combustible-fueled steam electric generator(s) and/or combined cycle steam generator(s) with duct
> firing.

e-belfer commented:
Outstanding issues to resolve:
- `boiler_fuel_5` - `boiler_fuel_8` are always empty. These trip up the no_null_columns() test. Should we get rid of these columns even though they exist in the original dataset, or how does PUDL deal with other always empty columns?

e-belfer commented:
Currently the full ETL works, tox is passing, and there is one outstanding issue with the validation tests:
- `bf_eia_923` now has ten records that have a duplicated primary key, with different fuel mmbtu / fuel consumed values.
Neither the boilers entity table nor boilers_eia860 have duplicates, so I'm not immediately sure where the duplicates are originating from. No other output tables fail the duplication test.

```
import pudl
import sqlalchemy as sa
pudl_settings = pudl.workspace.setup.get_defaults()
pudl_engine = sa.create_engine(pudl_settings['pudl_db'])
pudl_out = pudl.output.pudltabl.PudlTabl(
    pudl_engine=pudl_engine,
    fill_net_gen=True,
    freq="AS"
)
test = pudl_out.bf_eia923()
dupes = test.loc[test.duplicated(['plant_id_eia', 'report_date', 'boiler_id', 'energy_source_code'], keep = False), ['plant_id_eia', 'report_date', 'boiler_id']]
dupes

	plant_id_eia	report_date	boiler_id
7518	688	2008-01-01	2
7520	688	2008-01-01	2
21108	2059	2010-01-01	7
21117	2059	2010-01-01	7
23667	2336	2008-01-01	1
23668	2336	2008-01-01	1
29457	2963	2008-01-01	3302
29458	2963	2008-01-01	3302
38406	4939	2010-01-01	2
38410	4939	2010-01-01	2
76237	50482	2008-01-01	B1
76238	50482	2008-01-01	B1
77287	50625	2010-01-01	24
77288	50625	2010-01-01	24
77361	50625	2008-01-01	41
77363	50625	2008-01-01	41
112976	56309	2010-01-01	HRSG1
112990	56309	2010-01-01	HRSG1
112991	56309	2010-01-01	HRSG2
113005	56309	2010-01-01	HRSG2
```

zaneselvans commented:
Above you said:

> In boiler_fuel_eia923 table from which the boiler prime_mover_code originates, one boiler can have more than one prime_mover_code associated with it in any given year (one record per plant, boiler and energy source code). Given this information, presumably it wouldn't make sense to harvest this record into either a static or annual table which expects 1:1 boiler attributes, since we'd lose information on 1:m records?

Which would seem to indicate that the `prime_mover_code` in the `boiler_fuel_eia923` should be part of the primary key, and should be retained in the table (i.e. the fuel consumption is being reported on the basis of boiler, date, prime mover, and energy source code) Does that seem correct? If you include `prime_mover_code` in here and 

However, it's not listed in the table's primary key on your branch. **It's not listed in the resource definition at all on `dev`.** 🤦🏼 and I also get no duplicate records when I run your test on a fresh DB from `dev`.

If you look at the `pudl.transform.eia923.boiler_fuel()` function, we are aggregating 4.5% of the records to make them unique on the basis of the stated primary key (`report_date`, `plant_id_eia`, `boiler_id`, `energy_source_code`), and in the `_aggregate_duplicate_boiler_fuel_keys()` function there's a reference to `prime_mover_code`!!!

This is concerning. Have we been inappropriately ditching the `prime_mover_code` in this table all along?! @cmgosnell is there something I'm forgetting about this table transform?  It seems like this could have annoying implications for the fuel allocation process.

@e-belfer if you do the duplicate record check above again, but include `energy_source_code` and `prime_mover_code` in the outputs (and the primary key you're using to identify duplicates) what happens?

e-belfer commented:
`prime_mover_code` is not in `boiler_fuel_eia923` because it was getting pulling into the boilers entity table, which as we've discussed above, we'd like to change. So I've added it back to the `boiler_fuel_eia923` here. While before prime_mover_code was getting pulled out during the harvest and added back into the output table, it now lives in this table throughout. 

In the transformed boiler_fuel_eia923 table and at the end of the harvest, the boiler_fuel_eia923 has as a unique primary key `report_date, plant_id_eia, boiler_id, energy_source_code`. The issue here is actually occurring in the output table in `pudl.src.output.eia923.py`, line 478, where `prime_mover_code` is listed as one of the grouping entities for time-based aggregations. If we remove it the duplication issue is resolved without changing the primary key.

Alternately, we could add prime_mover_code to the primary key throughout, but I'm not sure what this would break. @zaneselvans or @cmgosnell you have thoughts on other ways to proceed or other places I should be updating this table, do let me know!

zaneselvans commented:
What I'm saying is that because there are multiple records with the same combination of (plant, boiler, energy source, date) but different prime movers, (plant, boiler, energy source, date) can't actually be the primary key for this table. The PK has to be unique. It seems like we brute forced our way around this by aggregating some of the records, after erroneously harvesting the prime mover codes, leaving the table with no valid primary key

If we:
* add prime mover back into the boiler fuel table (where it is now)
* make it part of the primary key, and
* stop aggregating records in the boiler fuel transform

Do we end up with a table that has unique natural primary keys?  If so, then I think this is probably the real structure of the table and we should probably use it and deal with the downstream consequences.

@cmgosnell what do you think about this?

e-belfer commented:
After earlier rounds of smoothing, in effect there are ten records here that report both ST and CA as the primary mover for a single boiler/energy source code within a given year. Repeating what was done in the transform in the output table would in effect just drop aggregate the other columns of data from the ten ST records but record the final prime_mover_code as CA. This feels consistent with what was done in #852, but perhaps we want to revisit that decision!

```

plant_id_eia | report_date | boiler_id | energy_source_code | prime_mover_code
688 | 2008-01-01 | 2 | NG | CA
688 | 2008-01-01 | 2 | NG | ST
2059 | 2010-01-01 | 7 | NG | CA
2059 | 2010-01-01 | 7 | NG | ST
2336 | 2008-01-01 | 1 | NG | CA
2336 | 2008-01-01 | 1 | NG | ST
2963 | 2008-01-01 | 3302 | NG | CA
2963 | 2008-01-01 | 3302 | NG | ST
4939 | 2010-01-01 | 2 | NG | CA
4939 | 2010-01-01 | 2 | NG | ST
50482 | 2008-01-01 | B1 | DFO | CA
50482 | 2008-01-01 | B1 | DFO | ST
50625 | 2010-01-01 | 24 | NG | CA
50625 | 2010-01-01 | 24 | NG | ST
50625 | 2008-01-01 | 41 | NG | CA
50625 | 2008-01-01 | 41 | NG | ST
56309 | 2010-01-01 | HRSG1 | NG | CA
56309 | 2010-01-01 | HRSG1 | NG | ST
56309 | 2010-01-01 | HRSG2 | NG | CA
56309 | 2010-01-01 | HRSG2 | NG | ST



```

cmgosnell commented:
I think we were incorrectly removing the pm codes from the boiler fuel table. I think what you've got going on right now is the right call (i.e. no longer associating pm codes with boilers and instead leaving the pms in the boiler fuel table).

on the downstream implications re the allocate net gen process: I don't think making this change will cause problems. We had previously been merging in the boiler pm code back into the boiler fuel table in the `pudl.output.eia923.boiler_fuel_eia923` output function via merging in the `boilers_entity_eia` table. So structurally the table going into the allocation process will look the same but with more information.

There are some steps in the allocaiton process that attempt to account for the fact that the pm's in the boiler fuel table do not perfectly match the generators. I think we will be able to strip that out with this change. I was planning on jumping into that model in this sprint anyway so unless this change breaks the allocation process (which I don't think it will). You can leave that process to me if you want.

e-belfer commented:
Options:
1) Carry on using the existing mapping function throughout the output process, which conforms the output table to match the existing format of the transformed boiler_fuel table. Make a separate issue to fix the primary key of this table, strip out all the mapping and see what breaks elsewhere. If we pick this option, this PR should basically be ready for final review/to merge in the next day or so, but we'll be kicking the can down the road re: boiler_fuel table (but not changing the status quo of PUDL's existing treatment of the PM codes).
2) Change the primary key for boiler_fuel in this PR and deal with these issues and whatever breaks in the allocate_net_gen() process, if anything, in this PR.
